draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -717,6 +774,175 @@ Steps:
+3. t1 = t0-B               // t1 = u^2-B
+4.  x = t1^c               // x  = (u^2-B)^(1/3)
+5. Output (x,u)
+~~~
+
+
+### Fouque-Tibouchi Method {#ftpairing}
+
+The map2curve_ft(alpha) implements the Fouque-Tibouchi's method {{FT12}} which
+covers the case of pairing-friendly curves `E: y^2=x^3+B`.
+Note that for pairing curves the destination group is usually a subgroup of the
+curve, hence, a scalar multiplication by the cofactor will be required to send
+the point to the desired subgroup.
+
+**Preconditions**
+
+This algorithm works for any Weierstrass curve over `Fq` such that `A=0`, `1+B` is
+a non-zero square in the field, and `q=7 mod 12`. This covers the case
+`q=1 mod 3` not handled by Boneh-Franklin's method.
+
+**Examples**
+
+- SECP256K1 curve {{SEC2}}
+- BN curves {{BN05}}
+- KSS curves {{KSS08}}
+- BLS curves {{BLS01}}
+
+**Algorithm**: map2curve_ft
+
+Input:
+
+ - `alpha`: an octet string to be hashed.
+ - `B`: the constant from the Weierstrass curve.
+ - `s`: a constant equal to sqrt(-3) in the field.
+
+Output:
+
+ - (x,y): a point in E.
+
+Operations:
+
+~~~
+1. t = HashToBase(alpha)
+2. w = (st)/(1+B+t^2)
+3. x1 = (-1+s)/2-tw
+4. x2 = -1-x1
+5. x3 = 1+1/w^2
+6. e = Legendre(t)
+7. if x1^3+B is square, output (x1, e * sqrt(x1^3+B) )
+8. if x2^3+B is square, output (x2, e * sqrt(x2^3+B) )
+9. Output (x3, e * sqrt(x3^3+B))
+~~~
+
+**Implementation**
+
+The following procedure implements the Fouque-Tibouchi's algorithm in a
+straight-line fashion.
+
+~~~
+map2curve_ft(alpha)
+
+Input:
+
+  alpha: an octet string to be encoded
+  B    : the constant of the curve
+
+Output:
+
+  (x, y): - a point in E
+
+Precomputations:
+
+1.  c1 = sqrt(-3)           // Field arithmetic
+2.  c2 = (-1+c1)/2          // Field arithmetic
+
+Steps:
+
+1.   u = HashToBase(alpha)  // {0,1}^* -> Fp
+2.  t0 = u^2                // u^2
+3.  t0 = t0+B+1             // u^2+B+1
+4.  t0 = 1/t0               // 1/(u^2+B+1)
+5.  t0 = t0*u               // u/(u^2+B+1)
+6.  t0 = t0*c1              // sqrt(-3)u/(u^2+B+1)
+7.  x1 = c2-u*t0            // (-1+sqrt(-3))/2-sqrt(-3)u^2/(u^2+B+1)
+8.  x2 = -1-x1
+9.  t1 = t0^2
+10. t1 = 1/t1
+11. x3 = t1+1
+12. fx1 = x1^3+B
+12. fx2 = x2^3+B
+12. s1 = Legendre(fx1)
+13. s2 = Legendre(fx2)
+14.  x = x3
+15.  x = CMOV(x2 ,x, s2>0)  // if s2=1, then x is set to x2
+16.  x = CMOV(x1, x, s1>0)  // if s1=1, then x is set to x1
+17.  y = x^3+B
+18. t2 = Legendre(u)
+19.  y = t2*sqrt(y)         // TODO: determine which root to choose
+20. Output (x, y)
+~~~
+
+Additionally, `map2curve_ft(alpha)` can return the point `(c2, sqrt(1+B))` when `u=0`.
+
+## Encodings for Montgomery curves
+
+A Montgomery curve is given by the following equation E: By^2=x^3+Ax^2+x, where
+B(A^2 − 4) ≠ 0. Note that any curve with a point of order 2 is isomorphic to
+this representation. Also notice that E cannot have a prime order group, hence,
+a scalar multiplication by the cofactor h={4,8} is required to obtain a point
```

Comment:
Can we remove the values for `h`, as those depend on the curve? We can just say that multiplication by the cofactor is required.

Reply:
Done